libraries/DateTime/contracts/DateTimeLibrary.aes

```diff
@@ -0,0 +1,430 @@
+    new_tampstamp
+
+  public function add_seconds(timestamp : int, seconds : int) : int = 
+    let new_tampstamp = timestamp + seconds
+    new_tampstamp
+
+  public function sub_seconds(timestamp : int, seconds : int) : int = 
+    let new_tampstamp = timestamp - seconds
+    new_tampstamp
+
+  public function is_leap_year(year : int) : bool =
+    if(rest(year, 4) != 0)
+      false
+    elif(rest(year, 100) != 0)
+      true
+    elif(rest(year, 400) != 0)
+      false
+    else
+      true
+
+  private function require(expression : bool, error : string) =
+    if(!expression) 
+      abort(error)
+
+  private function get_units_of_measure() : date_time_constants = {
+    origin_year = 1970, 
+    day_in_seconds = 86400,
+    year_in_seconds = 31536000,
+    leap_year_in_seconds = 31622400,
+    hour_in_seconds = 3600,
+    minute_in_seconds = 60 }
+
+  private function init_date_time(timestamp : int, constants : date_time_constants) : date_time =
+    let year = constants.origin_year + timestamp / constants.year_in_seconds
+    let total_leap_years_before = leap_years_before(year, constants.origin_year)
+    let seconds_accounted_for = constants.leap_year_in_seconds * total_leap_years_before
+    {
+      year = year,
+      month = 0,
+      day = 0,
+      hour = 0,
+      minute = 0,
+      second = 0,
+      weekday = 0,
+      leap_years_before = total_leap_years_before,
+      seconds_accounted_for = 
+        seconds_accounted_for + constants.year_in_seconds * 
+        (year - constants.origin_year - total_leap_years_before),
+      timestamp = timestamp
+      }
+
+  private function calculate_years_in_seconds(timestamp : int, dt : date_time) : int =
+    let constants : date_time_constants = get_units_of_measure()
+    let days : int = days_from_date(dt.year, dt.month, dt.day)
+    (days * constants.day_in_seconds) + timestamp mod constants.day_in_seconds
+
+  private function calculate_day(dt : date_time, days_in_month : int) : int =
+    if(dt.day > days_in_month)
+      let dt : date_time = dt { day = days_in_month }
+      dt.day
+    else
+      let dt : date_time = dt { day = dt.day }
+      dt.day
+  
+  // ------------------------------------------------------------------------
+  // Calculate year/month/day from the number of days since 1970/01/01 using
+  // the date conversion algorithm from
+  //   http://aa.usno.navy.mil/faq/docs/JD_Formula.php
+  // and adding the offset 2440588 so that 1970/01/01 is day 0
+  //
+  // int L = days + 68569 + offset
+  // int N = 4 * L / 146097
+  // L = L - (146097 * N + 3) / 4
+  // year = 4000 * (L + 1) / 1461001
+  // L = L - 1461 * year / 4 + 31
+  // month = 80 * L / 2447
+  // dd = L - 2447 * month / 80
+  // L = month / 11
+  // month = month + 2 - 12 * L
+  // year = 100 * (N - 49) + year + L
+  // ------------------------------------------------------------------------
+  private function days_to_date(days : int) : date_time = 
```

Comment:
lots of magic numbers in these ones

Reply:
They are explained above the functions. They are also private helper functions
taken elsewhere and i do not think there is much we can explain more than that

Reply:
I don't feel they are well explained, where can I find the explaination why e.g. `month = 80 * L / 2447`, I'm ok to add explainations as comments instead of resolving magic numbers in code